### ISSUE TITLE
Race condition in mkpath can lead to error

### ISSUE THREAD

**Issue by garrison** [JuliaLang/julia#13094]
Currently, `mkpath` checks for the existence of a path and then creates it if it does not exist.  However, if the path is brought into existence (e.g. by another process executing the same code) after the check but before the call to `mkdir`, `mkpath` will fail with a "File exists" error:

```
ERROR: LoadError: SystemError: mkdir: File exists
 in mkdir at file.jl:42
 in mkpath at file.jl:50
```

It would instead be better to catch this error and make sure that the "file" is indeed a directory, and simply continue rather than error out if it is.

**Comment by vtjnash** [comment]
technically, you can't completely avoid the race condition of someone else changing the filesystem after you checked (for example, some other process could equally have just deleted the directory), unless you were to also open a directory fd and then only ever use that to reference the filesystem (for example, by using the new `openat` system api's: http://linux.die.net/man/2/openat).

what does `mkdir -p` do? I thought i've seen the same race condition issue with it as well.

there does seem to be backing for making this change, based on similar patches existing in other large projects: http://stackoverflow.com/questions/19964287/mkdir-function-throw-exception-file-exists-even-after-checking-that-directory/25219407#25219407

**Comment by garrison** [comment]
> what does mkdir -p do? I thought i've seen the same race condition issue with it as well.

I think [you are correct](http://git.savannah.gnu.org/gitweb/?p=gnulib.git;a=blob;f=lib/mkdir-p.c;h=e55b1ddf145a5536055bee5cca9d188553c9f334;hb=HEAD), if I am reading things correctly.

**Comment by garrison** [comment]
In addition to the ones listed on StackOverflow, Python [is another project that has made this change](https://github.com/python/cpython/blob/829b49cbd2e4b1d573470da79ca844b730120f3d/Lib/os.py#L232-L234).

**Comment by StefanKarpinski** [comment]
Doing the try-catch thing seems pretty safe. I'm not sure what the downside would be.

**Comment by musicalrunner** [comment]
I'd like to work on this. I'm new to Julia development, but I think this is simple enough for me to tackle. Should I just implement the try-catch and file a pull request? It looks like test coverage for `mkpath` is pretty thorough, but I'm having trouble determining if there's one that tests this particular race condition.

**Comment by StefanKarpinski** [comment]
Yes, @musicalrunner, please do implement the try/catch thing. Once you've put a PR together, we can try to figure out a way to test it – race conditions are pretty hard to test, so I'm not really sure what we'll end up doing for that.

**Comment by musicalrunner** [comment]
This issue can probably be closed since the fix has been merged in (commit 3c8f45c)

**Comment by garrison** [comment]
Thanks!
